guides/md/writing_quantization_compatible_layers.md

```diff
@@ -0,0 +1,612 @@
+
+---
+## What are quantization-compatible layers?
+
+Keras lets you optimize models via post-training quantization (PTQ) by calling
+the `layer.quantize(...)` or `model.quantize(...)` APIs. Keras exposes an
+extensible framework for defining quantization-compatible layers. This lets you
+author custom layers that plug into the quantization framework, can be quantized
+to INT8 or INT4, and saved/loaded with quantization metadata.
+
+A quantization-compatible layer needs to implement a few hooks, so that it can:
+
+- Switch its variables to quantized representations.
+- Use a quantization-aware forward path at inference.
+- Save and load quantization metadata with the model.
+
+In this guide, we'll implement a simple layer that supports INT8 PTQ. The same
+patterns generalize to INT4 quantization and FP8 mixed-precision training.
+
+---
+## The hooks you'll implement
+
+At minimum, your layer should define:
+
+- `quantize(mode, **kwargs)`: Converts existing variables to quantized form and
+    switches the dtype policy
+- `_int8_build(...)`: Allocates INT8 variables needed by your layer
+- `_int8_call(inputs, training=None)`: Minimal INT8 forward path
+
+We'll implement these for a very small layer called `SimpleScale`, which
+multiplies the inputs by a trainable per-feature vector (elementwise scaling on
+the last dimension). The same patterns generalize to more sophisticated layers.
+
+---
+## Writing a Simple Quantization-Compatible Layer
+
+We start with a tiny layer that learns a per-feature multiplier. The
+full-precision path just computes `y = x * w`. We'll add the quantization hooks
+step by step.
+
+
+```python
+import numpy as np
+import keras
+from keras import ops, quantizers, dtype_policies
+from keras.layers import Layer, Input
+
+
+class SimpleScale(Layer):
+    """A layer that learns a per-feature scaling factor."""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self._kernel = self.add_weight(
+            name="kernel",
+            shape=(input_dim,),
+            initializer="random_uniform",
+        )
+
+    def call(self, inputs, training=None):
+        return ops.multiply(inputs, self._kernel)
+
+```
+
+### The `quantize()` method
+
+PTQ is a one-time rewrite. After you train or load your FP32 layer, you call
+`layer.quantize("int8")`. The layer should:
+
+1. Read its existing full-precision variables (e.g., `self._kernel`).
+2. Quantize them to INT8 values plus a quantization scale.
+3. Replace full-precision variables with INT8 storage and assign the quantized
+  data.
+4. Switch the `dtype_policy` to a quantized variant (e.g., `int8_from_float32`).
+
+
+```python
+
+def quantize(self, mode, **kwargs):
+    if mode != "int8":
+        raise NotImplementedError(f"Unsupported quantization mode: {mode}")
+
+    quantized_kernel, scale = quantizers.abs_max_quantize(
+        self._kernel, axis=0, to_numpy=True
+    )
```

Comment:
Can you explicitly pass `dtype="int8"` here? It's the default, but it makes you wonder how you would change this code to do "int4".